core/src/main/java/com/linecorp/armeria/common/util/SettableSupplier.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.client.limit.ConcurrencyLimit;
+
+/**
+ * Supplies the cached value that is useful below cases.
+ * <p>
+ * - Supplies dynamically a maximum number of concurrent active requests for {@link ConcurrencyLimit}.
+ * For example:
+ * <pre>{@code
+ * class DynamicSupplier extends SettableSupplier<Integer> {
+ *   public DynamicSupplier() {
+ *     super(10); //set initial value
+ *     AnyListener<Integer> listener = ...
+ *     listener.addListener(this::set);
+ *   }
+ * }}</pre>
+ */
+public class SettableSupplier<T> implements Supplier<T> {
+    private volatile T value;
```

Comment:
In my view, we can assume the value is rarely updated and read a lot.
Is it better to make this non-volatile and add a memory barrier?
```java
private int value;

private volatile boolean dirty;

public final void set(T value) {
    requireNonNull(value, "value");
    this.value = value;
    // forcibly flush the value
    dirty = true;
} 
```
You may want to run benchmarks to prove the performance.

Reply:
I misunderstood the happen-before relationship for volatile write. Please ignore this comment.  😅